code/game/objects/items/devices/radio/radio.dm

```diff
@@ -1,70 +1,108 @@
-	var/subspace_transmission = 0
-	var/syndie = 0//Holder to see if it's a syndicate encrypted radio
-	var/intercept = 0 //can intercept other channels
 	obj_flags = OBJ_FLAG_CONDUCTIBLE
 	slot_flags = SLOT_LOWER_BODY
 	throw_speed = 2
 	throw_range = 9
 	w_class = ITEM_SIZE_SMALL
-
-	material = /decl/material/solid/glass
-	matter = list(/decl/material/solid/metal/aluminium = MATTER_AMOUNT_REINFORCEMENT)
-	var/const/FREQ_LISTENING = 1
-	var/list/internal_channels
+	material = /decl/material/solid/metal/aluminium
+	matter = list(/decl/material/solid/glass = MATTER_AMOUNT_REINFORCEMENT)
 
 	var/obj/item/cell/cell = /obj/item/cell/device
 	var/power_usage = 2800
-
-	var/datum/radio_frequency/radio_connection
-	var/list/datum/radio_frequency/secure_radio_connections = new
-
 	var/last_radio_sound = -INFINITY
+	var/initial_network_id
+	var/initial_network_key
+	var/on = 1 // 0 for off
+	var/frequency = PUB_FREQ
+	var/traitor_frequency = 0 //tune to frequency to unlock traitor supplies
+	var/canhear_range = 3 // the range which mobs can hear this radio from
+	var/datum/wires/radio/wires = null
+	var/panel_open = 0
+	var/broadcasting = 0
+	var/listening = 1
+	var/can_use_peer_to_peer = TRUE
+	var/peer_to_peer_password
+	var/peer_to_peer = FALSE
+	var/peer_to_peer_range = 100
+	var/list/channels
+	var/default_color = "#6d3f40"
+	var/list/encryption_keys
+	var/encryption_key_capacity
+	var/decrypt_all_messages = FALSE
+
+/obj/item/radio/proc/can_decrypt(var/secured)
+	if(decrypt_all_messages)
+		return TRUE
+	if(!secured)
+		return TRUE
+	for(var/obj/item/encryptionkey/key in encryption_keys)
+		if(secured in key.can_decrypt)
+			return TRUE
+	return FALSE
 
 /obj/item/radio/proc/set_frequency(new_frequency)
-	radio_controller.remove_object(src, frequency)
-	frequency = new_frequency
-	radio_connection = radio_controller.add_object(src, frequency, RADIO_CHAT)
+	if(frequency != new_frequency)
+		frequency = new_frequency
 
 /obj/item/radio/Initialize()
 	. = ..()
 	wires = new(src)
 	if(ispath(cell))
 		cell = new(src)
-	internal_channels = GLOB.using_map.default_internal_channels()
-	GLOB.listening_objects += src
 
-	if(frequency < RADIO_LOW_FREQ || frequency > RADIO_HIGH_FREQ)
-		frequency = sanitize_frequency(frequency, RADIO_LOW_FREQ, RADIO_HIGH_FREQ)
-	set_frequency(frequency)
-
-	for (var/ch_name in channels)
-		secure_radio_connections[ch_name] = radio_controller.add_object(src, radiochannels[ch_name],  RADIO_CHAT)
+	GLOB.listening_objects += src
+	set_frequency(sanitize_frequency(frequency, RADIO_LOW_FREQ, RADIO_HIGH_FREQ))
+	set_extension(src, /datum/extension/network_device, initial_network_id, initial_network_key, NETWORK_CONNECTION_WIRELESS)
+
+	var/list/created_encryption_keys
+	for(var/keytype in encryption_keys)
+		LAZYADD(created_encryption_keys, new keytype(src))
+	encryption_keys = created_encryption_keys
+
+/obj/item/radio/proc/get_available_channels()
+	if(!channels)
+		sync_channels_with_network()
+	return channels
```

Comment:
all this caching is spooking me, need to make sure cache invalidates properly, like when you change network, maybe give a manual button in UI to refresh channel list.

Reply:
They have one.